shell/server/core.js

```diff
@@ -0,0 +1,220 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2014 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+var Crypto = Npm.require("crypto");
+var Capnp = Npm.require("capnp");
+
+var HandlePersistent = Capnp.importSystem("sandstorm/supervisor.capnp").HandlePersistent;
+var SandstormCore = Capnp.importSystem("sandstorm/supervisor.capnp").SandstormCore;
+var SandstormCoreFactory = Capnp.importSystem("sandstorm/backend.capnp").SandstormCoreFactory;
+var OngoingNotificationPersistent = Capnp.importSystem("sandstorm/supervisor.capnp").OngoingNotificationPersistent;
+
+function SandstormCoreImpl(grainId) {
+  this.grainId = grainId;
+}
+
+var makeSandstormCore = function (grainId) {
+  return new Capnp.Capability(new SandstormCoreImpl(grainId), SandstormCore);
+};
+
+function NotificationHandle(notificationId) {
+  this.notificationId = notificationId;
+}
+
+function makeNotificationHandle(notificationId) {
+  return new Capnp.Capability(new NotificationHandle(notificationId), HandlePersistent);
+}
+
+function dropWakelock(grainId, wakeLockNotificationId) {
+  // For some reason, Mongo returns an object that looks buffer-like, but isn't a buffer.
+  // We have to explicitly copy it before we can pass it on to node-capnp.
+  var copiedId = new Buffer(wakeLockNotificationId);
+  waitPromise(openGrain(grainId).supervisor.drop({wakeLockNotification: copiedId}).catch(function (err) {
+    if (shouldRestartGrain(err, 0)) {
+      return openGrain(grainId, true).supervisor.drop({wakeLockNotification: copiedId});
+    }
```

Comment:
You should re-throw the error if shouldRestartGrain() returns false (in all the places where you're calling shouldRestartGrain()).

Maybe we should have a helper method for this common pattern.

    useGrain(grainId, function (supervisor) {
      // do stuff with supervisor; whole function may be retried on disconnect
    });